src/r_builtins/paste.rs

```diff
@@ -0,0 +1,360 @@
+use crate::ast::*;
+use crate::error::*;
+use crate::lang::*;
+use crate::r_vector::vectors::*;
+
+pub fn primitive_paste(args: ExprList, env: &mut Environment) -> EvalResult {
+    // Need to make sure that collapse is specified since we can not assign a
+    // default value as with sep
+    let collapse_is_specified = &args
+        .keys
+        .iter()
+        .position(|k| k == &Some("collapse".to_string()));
+
+    let R::List(vals) = env.eval_list(args)? else {
+        unreachable!()
+    };
+
+    let mut stack_vals: Vec<(_, _)> = vec![];
+    let mut sep = " ".to_string();
+    let mut collapse = String::new();
+
+    for (k, v) in &vals {
+        let k_clone = k.clone().unwrap_or("".to_string());
+        let _sep_string = String::from("sep");
+
+        match k_clone.as_str() {
+            "sep" => {
+                sep = match v {
+                    R::Vector(Vector::Character(s_v)) => s_v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "sep parameter must be a character string!".to_string(),
+                        )))
+                    }
+                }
+            }
+            "collapse" => {
+                collapse = match v {
+                    R::Vector(Vector::Character(s_v)) => s_v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "collapse parameter must be a character string!".to_string(),
+                        )))
+                    }
+                }
+            }
+            _ => stack_vals.push((k, v)),
```

Comment:
There are _a lot_ of ways we could implement this, and I think this already gets us pretty far. I'll leave it up to you how far you want to take it, but I'll just mention what I see as areas for improvement. 

1. **Minimizing allocations**
    This code currently allocates a lot of intermediate vectors. Are these really needed? 
    In fact, if we know the length of all the vectors to be pasted together, then we should be able to determine the output length upfront. Furthermore, we shouldn't need to reallocate anything to build the new strings - we can just refer back to the input values. 
2. **Preallocate string buffer for collapse**
    In the case when collapse is non-NULL, we _should_ be able to calculate the total output length! We would then _only_ require that we allocate a new string with the appropriate length and populate it with the relevant pieces.

Reply:
Fully agree Doug and I excluded all the unnecessary extra loops. In essence we now have one loop to deal with function arguments and we do a lot of things in that one loop. The other loop is the main logic of paste with recycling and formatting the string.

The collapse is done with rust's join method. Do you think is not appropriate to join strings?

Reply:
This already covered with Doug's nice refactor.